File: frontend/src/types/invoice.ts
export interface LineItem {
  description: string;
  hsn_sac: string;
  gst_rate: number;
  quantity: number;
  rate: number;
}

export interface LineItemComputed extends LineItem {
  amount: number;
  gst_amount: number;
  total: number;
}

export interface ClientSnapshot {
  name: string;
  address: string;
  gstin: string;
  pan: string;
  email: string;
  phone: string;
  state: string;
}

export type InvoiceStatus = "unpaid" | "partial" | "paid";

export interface Invoice {
  id: string;
  invoice_no: string;
  invoice_date: string;
  due_date: string;
  client_id: string;
  client_snapshot: ClientSnapshot;
  line_items: LineItemComputed[];
  tax_type: "CGST_SGST" | "IGST";
  subtotal: number;
  cgst_total: number;
  sgst_total: number;
  igst_total: number;
  grand_total: number;
  gst_ratio: number;
  parent_id: string | null;
  remaining_line_items: LineItemComputed[] | null;
  paid_total: number;
  balance: number;
  status: InvoiceStatus;
  created_at: string;
  updated_at: string;
}

export interface InvoiceCreateInput {
  invoice_no: string;
  invoice_date: string;
  due_date: string;
  client_id: string;
  tax_type?: "CGST_SGST" | "IGST" | null;
  line_items: LineItem[];
}

File: backend/app/models/invoice.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel


class LineItem(BaseModel):
    description: str
    hsn_sac: str
    gst_rate: float
    quantity: float
    rate: float


class LineItemComputed(LineItem):
    amount: float
    gst_amount: float
    total: float


class ClientSnapshot(BaseModel):
    name: str
    address: str
    gstin: str
    pan: str
    email: str
    phone: str
    state: str


class InvoiceCreate(BaseModel):
    invoice_no: str
    invoice_date: date
    due_date: date
    client_id: str
    tax_type: Literal["CGST_SGST", "IGST"] | None = None
    line_items: list[LineItem]


class InvoiceResponse(BaseModel):
    id: str
    invoice_no: str
    invoice_date: date
    due_date: date
    client_id: str
    client_snapshot: ClientSnapshot
    line_items: list[LineItemComputed]
    tax_type: Literal["CGST_SGST", "IGST"]
    subtotal: float
    cgst_total: float
    sgst_total: float
    igst_total: float
    grand_total: float
    gst_ratio: float
    parent_id: str | None
    remaining_line_items: list[LineItemComputed] | None
    paid_total: float
    balance: float
    status: Literal["unpaid", "partial", "paid"]
    created_at: datetime
    updated_at: datetime
